frodo.js

```diff
@@ -1,119 +1,111 @@
-/**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?